service.py

```diff
@@ -0,0 +1,228 @@
+            type = filters.split('/')[0]
+            direction = filters.split('/')[1]
+            if direction == 'False':
+                direction = False
+            else:
+                direction = True
+            print(f'$$$$$$$$$$ SORT BY: {type} $$$$$$$')
+            print(f'$$$$$$$$$$ REVERSE: {direction} $$$$$$$')
+            card_list.sort(key=lambda x: x.get(f'{type}'), reverse=direction)
+            return card_list
+
+# Get users (top 5) with the most number of cards
+def get_leaderboard():
+    leaderboard = sql_select('SELECT username, count(card_id) FROM users INNER JOIN achievements AS ACH ON users.user_id = ACH.user_id GROUP BY username ORDER BY count(card_id) DESC LIMIT 5')
+    list = []
+    position = 1
+    for leader in leaderboard:
+        list.append(
+            {   
+                'position': position,
+                'name': leader[0].capitalize(),
+                'cardcount': leader[1]
+            }
+        )
+        position += 1
+    return list
+
+# Populate secret question options
+def get_questions(pos=None):
+    questions = {
+        1: 'What was your childhood pets name?',
+        2: 'What is your Fathers middle name?',
+        3: 'What is your Mothers maiden name?',
+        4: 'What is your favourite movie?',
+        5: 'What is your favourite food?',
+        6: 'What was your primary school name?',
+    }
+    if pos != None:
+        return questions[pos]
+    else: 
+        return questions
+
+# Main function to gather data from API and return focus hero + random choices
+def fetch_data():
+    excluded_letters = get_excluded_letters()    
+    letter = generate_random_letter(excluded_letters) 
+
+    url = 'http://gateway.marvel.com/v1/public/characters'
+    payload = get_payload(letter)
+
+    response = requests.get(url, params=payload).json()
+    attribute = response['attributionText'] 
+    characters = response['data']['results']
+
+    dict = update_dict(characters)            
+    if not dict:
+        return handle_excluded_letters(excluded_letters, letter)
+
+    selection = select_character(dict)     
+    choices = create_buttons(selection, characters)    
+    return selection, attribute, choices
+
+# Retrieve users 'excluded letters' as a string
+def get_excluded_letters():
+    query = sql_select('SELECT excluded FROM users WHERE user_id=%s', session['user_id'])
+    excluded_letters = query[0][0]
+    print(f'$$$$$$$$$ EXCLUDED LETTERS: {excluded_letters} $$$$$$$$$')
+    return excluded_letters
+
+# Generate a random letter that isnt already in the excluded list
+def generate_random_letter(excluded_letters):
+    letter = chr(randint(ord('A'), ord('Z')))
+    while letter in excluded_letters:
+        letter = chr(randint(ord('A'), ord('Z')))
+    return letter
+
+# Prepare payload for API query
+def get_payload(letter):
+    public = 'd609fc87a5b2e7b633e37e0e4cdf5553'
+    private = '082dff72dc33db36fd9194479cc71a83b9cf62d9'
```

Comment:
You probably don't want this in the code. There would be ways to have this set in the deployment of  your service. E.g. now I could use this private key. Not a big deal now but this can cost companies thousands of dollars in abuse.